### PR TITLE
docs: Runtime Defined Entity Behaviors Part 1

### PR DESCRIPTION
**Pull request by kirilkadoncheva**
This commit contains Behaviors general concepts, invocation, RDE hooks documentation.
    
Signed-off-by: Kirilka Doncheva <donchevak@vmware.com>

### CODE REVIEW

File: documentation/images/rde-async-multi-stage-delete.png
Comment (iiliev2) [inline documentation/images/rde-async-multi-stage-delete.png:1]: These images do not look well on light theme(there is no background and some text is white)

File: documentation/extensibility-platform/behaviors-general-concepts.md [hunk 1]
```diff
@@ -0,0 +1,549 @@
...
+  ...
+    "result": null,
+    "status": "success",
+    "operation": "PreDelete hook: urn:vcloud:task:7d0f585f-75ea-4af1-a94c-5a0727124d8f.", // pre-delete hook task
+    "operationName": "updateDefinedEntity",
+    ...
+}
+```
+
+In case the entity is already in ```IN_DELETION``` state when the PUT api call is made, the pre-delete hook will not be executed again, because we assume that the execution is successful.
```
Comment (iiliev2) [inline documentation/extensibility-platform/behaviors-general-concepts.md:539]: It is hard to follow this. IMO we should have more detailed state machine diagrams.

**Comment by kirilkadoncheva** [comment]
> Overall the last section with the hooks is rather unclear for me. The information should be regrouped. Always think from the point of view of a third party who do not know any of this:
> 
> * is the necessary context explained prior to any given new statement
> * there should be no duplicated information(also because when there are slight differences between the two versions it becomes even more confusing as to which one is the correct one)
> 
> If there will be so many documents for the RDEs we should put them in a sub folder.

I have added the RDE related documentation in a subfolder. Also reorganized the hooks section of the document.

File: documentation/extensibility-platform/behaviors-general-concepts.md [hunk 2]
```diff
@@ -0,0 +1,549 @@
...
+
+In the async scenario, the solution backend is expected to poll the Cloud Director API for entities which are marked for deletion.
+
+To put it simply, the async multi-stage deletion involves the following steps:
+
+1. Marking entities for deletion by moving them to ```IN_DELETION``` state (more info [here](#moving-entities-to-in_deletion-state)).
+2. The solution backend polls for entities in ```IN_DELETION``` state and starts the deletion process for them (more info [here](#polling-for-entities-to-in_deletion-state)).
+3. Once resource clean-up is completed, solution backend can issue a ```DELETE``` API call for all entities, which are ready to be permanently deleted.
+
+![RDE async multi-stage delete diagram](../images/rde-async-multi-stage-delete.png)
```
[the remark this answers is not in the document]
Comment (kirilkadoncheva) [reply documentation/extensibility-platform/behaviors-general-concepts.md]: No, the entity is already in `IN_DELETION` state, so the pre-delete hook invocation will be skipped during `DELETE`. I added more diagrams to illustrate this

File: documentation/extensibility-platform/behaviors-general-concepts.md [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+```json
+"execution": {
+         "type": "...",
+         "_internal_key": "...",
+         "execution_properties": {
+             "_internal_key_1": "..."
+         }
+}
+```
+- fields with prefix ```_secure_``` - field is write-only - once it is set it cannot be obtained through a GET request on the behavior. The field value is saved encrypted in the DB. However, this field is accessible to behavior executors. This means a different thing in the context of different types of behaviors (i.e. these fields are accessible to the webHook behaviors' template). These fields work on top level in the behavior ```execution``` or ```execution_properties```.
```
[the remark this answers is not in the document]
Comment (kirilkadoncheva) [reply documentation/extensibility-platform/behaviors-general-concepts.md]: I removed this as a term.

File: documentation/extensibility-platform/behaviors-general-concepts.md [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+### Pre Delete & Post Delete Behavior Hooks (Multi-stage RDE Deletion)
+The pre-delete and post-delete hook behaviors are hooked to the RDE deletion operation. A multi-stage entity deletion process can be achieved using these hooks. 
+
+#### Pre Delete Hook Behavior
+The pre-delete hook behavior is invoked before the actual entity deletion happens. If the hook execution is successful, the entity deletion proceeds. However, if the hook execution fails, then the entity deletion is "canceled" - the entity remains unchanged. 
+
+The intended use of this hook is as a pre-check for whether the entity can be deleted depending on the extension logic.
+
+#### Post Delete Hook Behavior
+The pre-delete hook behavior is invoked during the actual entity deletion process. The entity is put in ```IN_DELETION``` first to signal that it is being deleted. Then the post-delete hook is invoked. If the hook execution is successful, the entity is permanently deleted. If the hook execution fails, the entity deletion process fails and entity remains in ```IN_DELETION``` state.
```
[the remark this answers is not in the document]
Comment (kirilkadoncheva) [reply documentation/extensibility-platform/behaviors-general-concepts.md]: I re-worked the hooks section

File: documentation/extensibility-platform/behaviors-general-concepts.md [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+When a defined entity with a post-create hook is created, the task returned in the ```Location``` response of the create RDE API call is the behavior invocation task.
+### Post Update Behavior Hook
+The post-update hook behavior is invoked automatically after a defined entity instance update. The behavior execution does not affect the lifecycle of the RDE instance in any way.
+
+When a defined entity with a post-update hook is updated, the behavior invocation task is returned in a  ```X-VMWARE-VCLOUD-TASK-LOCATION``` header in the response of the update RDE API call.
+### Pre Delete & Post Delete Behavior Hooks (Multi-stage RDE Deletion)
+The pre-delete and post-delete hook behaviors are hooked to the RDE deletion operation. A multi-stage entity deletion process can be achieved using these hooks. 
+
+#### Pre Delete Hook Behavior
+The pre-delete hook behavior is invoked before the actual entity deletion happens. If the hook execution is successful, the entity deletion proceeds. However, if the hook execution fails, then the entity deletion is "canceled" - the entity remains unchanged.
```
[the remark this answers is not in the document]
Comment (kirilkadoncheva) [reply documentation/extensibility-platform/behaviors-general-concepts.md]: I re-worded this and added more diagrams

File: documentation/extensibility-platform/defined-entities/behaviors-general-concepts.md [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+    "ref": "urn:vcloud:behavior-interface:noopBehavior:vmware:test:1.0.0",
+    "description": null,
+    "execution": {
+        "type": "noop"
+    }
+}
+```
+
+### Type Behaviors
+Each RDE Type implements one or more interfaces. It is possible to override the execution and description of each interface behavior (in the implemented interfaces) in the RDE type:
```
Comment (theotonchev) [inline documentation/extensibility-platform/defined-entities/behaviors-general-concepts.md:49]: Each RDE Type can implements one or more interfaces. The behaviors defined in those interfaces can be executed on instances of the RDE Type.

It is possible for the RDE Type definition to override the execution and description of interface behaviors. If an overridden behavior is executed on an instance of that RDE Type, then the behavior defined in the RDE Type will be invoked, rather than the default behavior in the inherited Interface, even if the behavior is specified by its Interface ID.

File: documentation/extensibility-platform/defined-entities/behaviors-general-concepts.md [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+        "href": "...",
+        "type": "WebHook",
+        "execution_properties": {
+            "actAsToken": true
+        }
+    }
+}
+```
+
+When executing a behavior on a RDE instance, if that behavior is overridden in the RDE type, the Type Behavior will be executed (even if the behavior is invoked by `ref`).
```
Comment (theotonchev) [inline documentation/extensibility-platform/defined-entities/behaviors-general-concepts.md:89]: Moved above

File: documentation/extensibility-platform/defined-entities/behaviors-general-concepts.md [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+    "id": "urn:vcloud:behavior-type:behaviorName:typeVendor:typeName:typeVersion:interfaceVendor:interfaceName:interfaceVersion",
+    "ref": "urn:vcloud:behavior-interface:behaviorName:interfaceVendor:interfaceName:interfaceVersion",
+    "description": null,
+    "execution": {
+        "type": "...",
+        "execution_properties": {...}
+    }
+}
+```
+The behavior __execution type__ is set in the `type` field of the `execution`.
```
Comment (theotonchev) [inline documentation/extensibility-platform/defined-entities/behaviors-general-concepts.md:105]: nit: of the execution section

File: documentation/extensibility-platform/defined-entities/behaviors-general-concepts.md [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+The `id` property holds the actual behavior id. If the behavior is not overridden, it is a `behavior-interface` id: 
+```
+urn:vcloud:behavior-interface:behaviorName:interfaceVendor:interfaceName:interfaceVersion
+```
+Otherwise, it is a `behavior-type` id: 
+```
+urn:vcloud:behavior-type:behaviorName:typeVendor:typeName:typeVersion:interfaceVendor:interfaceName:interfaceVersion
+```
+
+The `ref` property always holds a reference to the interface behavior to be used for polymorphic behavior invocations - it always holds the `behavior-interface` id.
```
Comment (theotonchev) [inline documentation/extensibility-platform/defined-entities/behaviors-general-concepts.md:118]: A behavior representation has both an `id` and a `ref` fields. 

The id property holds the actual behavior id. It will be an interface behavior id if the behavior is defined in an interface and it will be a type behavior id if the behavior is defined in a type and overrides an interface behavior.

The ref property holds a reference to the interface behavior definition that this behavior implements. It is used for polymorphic behavior invocations.

If the behavior is an interface behavior, then the id and the ref properties are identical.
If the behavior is a type behavior, however, then the id property contains the id of the type behavior and the ref property contains the id of the interface behavior that has been overridden, for example:
```json
    "id": "urn:vcloud:behavior-type:behaviorName:typeVendor:typeName:typeVersion:interfaceVendor:interfaceName:interfaceVersion",
    "ref": "urn:vcloud:behavior-interface:behaviorName:interfaceVendor:interfaceName:interfaceVersion",
```

File: documentation/extensibility-platform/defined-entities/behaviors-general-concepts.md [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+
+#### Special execution properties
+There are some special properties which can be set in the behavior definition `execution` or `execution_properties`.
```
Comment (theotonchev) [inline documentation/extensibility-platform/defined-entities/behaviors-general-concepts.md:121]: There are some special properties that can be set in the behavior definition `execution` or `execution_properties` sections.
...
If set to static, the behavior can be invoked
...
- fields with the prefix `_internal_` are write-only. Once they are set, they cannot be obtained through a GET request on the behavior. The field value is saved in the DB in an encrypted form. It is only accessible to Cloud Director (e.g. the shared secret in webHook behaviors). These fields can be defined at the top level of the behavior's `execution` or `execution_properties` sections.

- fields with the prefix `_secure_` are write-only. Once they are set, they cannot be obtained through a GET request on the behavior. The field value is saved in the DB in an encrypted form. Such fields are accessible to the behavior execution code, however. ...
These fields can be defined at the top level of the behavior's `execution` or `execution_properties` sections.

File: documentation/extensibility-platform/defined-entities/behaviors-general-concepts.md [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+    }
+}
+```
+
+## Behavior Invocation
+Behavior invocation is an asynchronous operation in Cloud Director since a behavior execution is a long running process. For each behavior invocation, a `BEHAVIOR_INVOCATION` task is created to track the execution. 
+
+Behaviors can be defined as either dynamic or static.
+
+The invocation of dynamic behaviors applies only to their existence on a defined entity - they need an RDE instance to be invoked on.
```
Comment (theotonchev) [inline documentation/extensibility-platform/defined-entities/behaviors-general-concepts.md:189]: Dynamic behaviors must be invoked in the context of an RDE instance.

File: documentation/extensibility-platform/defined-entities/behaviors-general-concepts.md [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+
+Static (or standalone) behaviors do not need to refer to a defined entity instance in order to execute. However, such behaviors can be executed on a defined entity instance as well.
```
Comment (theotonchev) [inline documentation/extensibility-platform/defined-entities/behaviors-general-concepts.md:191]: Static (or standalone) behaviors do not need to refer to a defined entity instance in order to be invoked. However, such behaviors can be invoked in the context of a defined entity instance as well.

File: documentation/extensibility-platform/defined-entities/behaviors-general-concepts.md [hunk 13]
```diff
@@ -0,0 +1,410 @@
...
+POST /cloudapi/1.0.0/entities/<entity-id>/behaviors/<behavior-id>/invocations
+```
+```json
+{
+    "arguments": {
+        "x": 7
+    }, 
+    "metadata": {
+        "y": 6
+    }
```
Comment (theotonchev) [inline documentation/extensibility-platform/defined-entities/behaviors-general-concepts.md:208]: We need a better example for invocation metadata here. 
Perhaps "convert": true ?

File: documentation/extensibility-platform/defined-entities/behaviors-general-concepts.md [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+        "x": 7,
+        ...
+    }, 
+    "metadata": {
+        "y": 6,
+        ...
+    } 
+}
+```
+#### Payload Behavior Execution Code Recieves
```
Comment (theotonchev) [inline documentation/extensibility-platform/defined-entities/behaviors-general-concepts.md:252]: Payload Received By the Behavior Execution Code

File: documentation/extensibility-platform/behaviors-general-concepts.md [hunk 15]
```diff
@@ -0,0 +1,549 @@
...
+- ```urn:vcloud:accessLevel:ReadWrite```- if ```accessLevelId``` is set to this value, an API user must have at least RW (read-write) access to a defined entity instance in order to invoke the behavior with id ```<behavior-id>``` on that defined entity instance
+- ```urn:vcloud:accessLevel:FullControl```- if ```accessLevelId``` is set to this value, an API user must have FC (full control) access to a defined entity instance in order to invoke the behavior with id ```<behavior-id>``` on that defined entity instance
+
+More information on RDE access control can be found [here](). <!-- TODO add link -->
+
+### Static Behaviors Access Control
+Currently, static behaviors do not have an access control mechanism for execution.
+
+## Runtime Defined Entity Lifecycle Hooks
+Behaviors can be configured to be triggered by the following Defined Entity Events:
```
[the remark this answers is not in the document]
Comment (kirilkadoncheva) [reply documentation/extensibility-platform/behaviors-general-concepts.md]: I re-worked the hooks section but it will be part of the RDE, not Behavior, documentation.

**Review by iiliev2 — APPROVED** [review]

**Review by theotonchev — APPROVED** [review]
